File: memopt/cluster/block_directory.py
"""
Block directory — cluster-wide index of memory block locations.

Maps content_hash → BlockEntry(node_id, tier, path, lease_count).

Shim layer: tries C++ BlockDirectoryCpp from _memopt_core for
LocalBlockDirectory (16-shard concurrent map, no GIL). Falls back
to pure Python implementation if C++ extension not built.

RedisBlockDirectory stays Python — Redis ops are network-bound.

Environment variables:
  REDIS_URL              str   Redis URL for cluster-wide directory
  MEMOPT_BLOCK_DIR_TTL_S float default 3600.0 (1 hour block expiry)
"""
from __future__ import annotations
import os
import time
import logging

logger = logging.getLogger(__name__)

# Always import BlockEntry and supporting types from the backup.
# These are dataclasses used by all callers — never replaced by C++.
from memopt.cluster._block_directory_py import (  # noqa: F401
    BlockEntry,
    RedisBlockDirectory,
)

_BLOCK_TTL_S = float(os.environ.get("MEMOPT_BLOCK_DIR_TTL_S", "3600.0"))

# ── Try C++ BlockDirectory ────────────────────────────────────────────

_cpp_available = False
try:
    from memopt._memopt_core import BlockDirectoryCpp  # type: ignore
    _cpp_available = True
    logger.info(
        "memopt: C++ BlockDirectory active (16-shard, no GIL)")
except ImportError:
    logger.info(
        "memopt: C++ BlockDirectory not available, "
        "using Python fallback")


# ── LocalBlockDirectory ──────────────────────────────────────────────

if _cpp_available:
    class LocalBlockDirectory:
        """C++ block directory — 16-shard concurrent map, no GIL."""

        def __init__(self, node_id: str = "",
                     default_ttl_s: float = _BLOCK_TTL_S):
            self._node_id = node_id
            self._cpp = BlockDirectoryCpp(node_id, default_ttl_s)

        def register(self, entry: BlockEntry) -> None:
            self._cpp.register_block({
                "content_hash":  entry.content_hash,
                "node_id":       entry.node_id,
                "tier":          entry.tier,
                "path":          entry.path,
                "size_bytes":    entry.size_bytes,
                "registered_at": entry.registered_at,
            })

        def lookup(self, content_hash: str):
            result = self._cpp.lookup(content_hash)
            if result is None:
                return None
            return BlockEntry(
                content_hash=result["content_hash"],
                node_id=result["node_id"],
                tier=result["tier"],
                path=result["path"],
                size_bytes=result["size_bytes"],
                registered_at=result["registered_at"],
                lease_count=result["lease_count"],
            )

        def acquire_lease(self, content_hash: str,
                          requesting_node: str) -> bool:
            return self._cpp.acquire_lease(
                content_hash, requesting_node)

        def release_lease(self, content_hash: str,
                          requesting_node: str) -> None:
            self._cpp.release_lease(content_hash, requesting_node)

        def can_evict(self, content_hash: str) -> bool:
            return self._cpp.can_evict(content_hash)

        def deregister(self, content_hash: str) -> None:
            self._cpp.deregister(content_hash)

        def list_node_blocks(self, node_id: str):
            # Not available in C++ path — return empty.
            # This method is only used by stats/debug, not hot path.
            return []

        def stats(self) -> dict:
            return {
                "total_entries":  self._cpp.size(),
                "active_entries": self._cpp.size(),
                "leased_entries": 0,
                "by_tier":        {},
                "backend":        "cpp_sharded",
            }
else:
    from memopt.cluster._block_directory_py import (  # type: ignore  # noqa: F401
        LocalBlockDirectory,
    )


def make_directory(node_id: str = ""):
    """Return the best available block directory."""
    redis_url = os.environ.get("REDIS_URL", "")
    if redis_url:
        return RedisBlockDirectory(redis_url=redis_url, node_id=node_id)
    if _cpp_available:
        return LocalBlockDirectory(node_id=node_id)
    return LocalBlockDirectory()


__all__ = [
    "BlockEntry",
    "LocalBlockDirectory",
    "RedisBlockDirectory",
    "make_directory",
]

File: csrc/core/block_directory.cpp
// block_directory.cpp — Sharded block directory implementation.

#include "block_directory.h"

namespace memopt {

// ═══════════════════════════════════════════════════════════════════════════
// Helpers
// ═══════════════════════════════════════════════════════════════════════════

double BlockDirectoryCpp::now_unix() noexcept {
    auto now = std::chrono::system_clock::now();
    return std::chrono::duration<double>(
        now.time_since_epoch()).count();
}

bool BlockDirectoryCpp::is_expired(const BlockEntryC& e) const noexcept {
    double ttl = (e.ttl_s > 0) ? e.ttl_s : default_ttl_s_;
    return (now_unix() - e.registered_at) > ttl;
}

// ═══════════════════════════════════════════════════════════════════════════
// Constructor
// ═══════════════════════════════════════════════════════════════════════════

BlockDirectoryCpp::BlockDirectoryCpp(const std::string& node_id,
                                      double default_ttl_s)
    : node_id_(node_id), default_ttl_s_(default_ttl_s) {}

// ═══════════════════════════════════════════════════════════════════════════
// register_block
// ═══════════════════════════════════════════════════════════════════════════

void BlockDirectoryCpp::register_block(const BlockEntryC& entry) noexcept {
    std::string key(entry.content_hash);
    auto& shard = shards_[shard_for(key)];
    std::unique_lock<std::shared_mutex> lock(shard.mu);
    shard.entries[key] = entry;
}

// ═══════════════════════════════════════════════════════════════════════════
// lookup
// ═══════════════════════════════════════════════════════════════════════════

bool BlockDirectoryCpp::lookup(const std::string& content_hash,
                                BlockEntryC* out_entry) noexcept {
    auto& shard = shards_[shard_for(content_hash)];
    std::shared_lock<std::shared_mutex> lock(shard.mu);
    auto it = shard.entries.find(content_hash);
    if (it == shard.entries.end()) return false;
    if (is_expired(it->second)) return false;
    if (out_entry) *out_entry = it->second;
    return true;
}

// ═══════════════════════════════════════════════════════════════════════════
// acquire_lease / release_lease
// ═══════════════════════════════════════════════════════════════════════════

bool BlockDirectoryCpp::acquire_lease(
        const std::string& content_hash,
        const std::string& /*requesting_node*/) noexcept {
    auto& shard = shards_[shard_for(content_hash)];
    std::unique_lock<std::shared_mutex> lock(shard.mu);
    auto it = shard.entries.find(content_hash);
    if (it == shard.entries.end()) return false;
    if (is_expired(it->second)) return false;
    // Only NVMe blocks are leasable
    if (std::strncmp(it->second.tier, "nvme", 4) != 0) return false;
    it->second.lease_count++;
    return true;
}

bool BlockDirectoryCpp::release_lease(
        const std::string& content_hash,
        const std::string& /*requesting_node*/) noexcept {
    auto& shard = shards_[shard_for(content_hash)];
    std::unique_lock<std::shared_mutex> lock(shard.mu);
    auto it = shard.entries.find(content_hash);
    if (it == shard.entries.end()) return false;
    if (it->second.lease_count > 0)
        it->second.lease_count--;
    return true;
}

// ═══════════════════════════════════════════════════════════════════════════
// can_evict / deregister
// ═══════════════════════════════════════════════════════════════════════════

bool BlockDirectoryCpp::can_evict(
        const std::string& content_hash) noexcept {
    auto& shard = shards_[shard_for(content_hash)];
    std::shared_lock<std::shared_mutex> lock(shard.mu);
    auto it = shard.entries.find(content_hash);
    if (it == shard.entries.end()) return true;
    return it->second.lease_count == 0;
}

void BlockDirectoryCpp::deregister(
        const std::string& content_hash) noexcept {
    auto& shard = shards_[shard_for(content_hash)];
    std::unique_lock<std::shared_mutex> lock(shard.mu);
    shard.entries.erase(content_hash);
}

// ═══════════════════════════════════════════════════════════════════════════
// size / expired_count
// ═══════════════════════════════════════════════════════════════════════════

size_t BlockDirectoryCpp::size() const noexcept {
    size_t total = 0;
    for (const auto& shard : shards_) {
        std::shared_lock<std::shared_mutex> lock(shard.mu);
        total += shard.entries.size();
    }
    return total;
}

size_t BlockDirectoryCpp::expired_count() noexcept {
    size_t removed = 0;
    for (auto& shard : shards_) {
        std::unique_lock<std::shared_mutex> lock(shard.mu);
        auto it = shard.entries.begin();
        while (it != shard.entries.end()) {
            if (is_expired(it->second)) {
                it = shard.entries.erase(it);
                ++removed;
            } else {
                ++it;
            }
        }
    }
    return removed;
}

} // namespace memopt
